### PR TITLE
DAOS-12710 build: Improve daos/tse.h handling.

### PR DESCRIPTION
**Pull request by ashleypittman**
Fix doxygen comments in daos_task.h and daos/tse.h

Change includes so all public daos headers are independent and can
be included without including any headers prior.
Add a unit test about the above for all installed headers.
Remove the scons check for gurt and cart headers in favour
of the new unit test.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
@mchaarawi this is technically an API change but did we intend to export daos/tse.h?  If there are functions there we want exported then we should move them to daos_task.h IMO.

**Comment by mchaarawi** [comment]
> @mchaarawi this is technically an API change but did we intend to export daos/tse.h? If there are functions there we want exported then we should move them to daos_task.h IMO.

not really. daos_task is something different than tse. tse by itself is unrelated to DAOS and meant as a separate package. some libraries use tse without daos. but daos_task is different and depends on DAOS.

**Comment by ashleypittman** [comment]
> > @mchaarawi this is technically an API change but did we intend to export daos/tse.h? If there are functions there we want exported then we should move them to daos_task.h IMO.
> 
> not really. daos_task is something different than tse. tse by itself is unrelated to DAOS and meant as a separate package. some libraries use tse without daos. but daos_task is different and depends on DAOS.

Does anything outside the daos tree use tse directly?

With this PR you can't include daos/tse.h without previously including daos_task.h which goes against what you're saying but I suspect that was the case anyway - I can check.

**Review by mchaarawi — CHANGES_REQUESTED** [review]
tse is used by HDF5 and possible other packages, so this cannot land.

**Comment by mchaarawi** [comment]
> > > @mchaarawi this is technically an API change but did we intend to export daos/tse.h? If there are functions there we want exported then we should move them to daos_task.h IMO.
> > 
> > 
> > not really. daos_task is something different than tse. tse by itself is unrelated to DAOS and meant as a separate package. some libraries use tse without daos. but daos_task is different and depends on DAOS.
> 
> Does anything outside the daos tree use tse directly?
> 
> With this PR you can't include daos/tse.h without previously including daos_task.h which goes against what you're saying but I suspect that was the case anyway - I can check.

daos_task.h depends on tse.h not the other way around. so this PR is not correct.

**Comment by ashleypittman** [comment]
> tse is used by HDF5 and possible other packages, so this cannot land.

Ok, updated on that basis.  daos/tse.h was and still is includable without pulling in any daos_*.h headers.

File: src/common/tse.c [hunk 1]
```diff
@@ -17,8 +17,10 @@
 
 #include <stdint.h>
 #include <pthread.h>
+#include <daos_task.h>
```
Comment (mchaarawi) [inline src/common/tse.c:20]: where does this file require daos_task.h ?

File: src/include/daos/common.h [hunk 2]
```diff
@@ -25,17 +25,21 @@
...
 #include <daos_errno.h>
-#include <daos/debug.h>
-#include <gurt/hash.h>
-#include <gurt/common.h>
-#include <cart/api.h>
 #include <daos_types.h>
 #include <daos_obj.h>
 #include <daos_prop.h>
 #include <daos_security.h>
+#include <daos_task.h>
```
Comment (mchaarawi) [inline src/include/daos/common.h:32]: why add daos_task.h here?

File: src/include/daos/common.h [hunk 3]
```diff
@@ -25,17 +25,21 @@
...
+
+#include <daos/debug.h>
+#include <gurt/hash.h>
+#include <gurt/common.h>
+#include <cart/api.h>
+
 #include <daos/profile.h>
 #include <daos/dtx.h>
 #include <daos/cmd_parser.h>
+#include <daos/tse.h>
```
Comment (mchaarawi) [inline src/include/daos/common.h:42]: same, why is that needed here?

File: src/include/daos/dpar.h [hunk 4]
```diff
@@ -9,10 +9,6 @@
 #include <stdbool.h>
 #include <inttypes.h>
 
-#ifdef __cplusplus
```
Comment (mchaarawi) [inline src/include/daos/dpar.h:12]: why remove those?

File: src/include/daos/tse.h [hunk 5]
```diff
@@ -15,18 +15,24 @@
...
-/* internal tse private data size (struct tse_task_private) */
+/** internal tse private data size (struct tse_task_private) */
 #define TSE_PRIV_SIZE		(TSE_TASK_ARG_LEN + 136)
-/* tse_task is used to track single asynchronous operation (8 bytes used for public members). */
+/** tse_task is used to track single asynchronous operation (8 bytes used for public members). */
 #define TSE_TASK_SIZE		(TSE_PRIV_SIZE + 8)
 
+/** A single task */
 typedef struct tse_task {
+	/** the result */
```
Comment (mchaarawi) [inline src/include/daos/tse.h:31]: that sounds like a useless comment.

File: src/include/daos/tse.h [hunk 6]
```diff
@@ -15,18 +15,24 @@
...
+/** tse_task arguments max length (pthread_mutex_t is of different size between x86 and aarch64). */
 #define TSE_TASK_ARG_LEN	(840 + sizeof(pthread_mutex_t))
-/* internal tse private data size (struct tse_task_private) */
+/** internal tse private data size (struct tse_task_private) */
 #define TSE_PRIV_SIZE		(TSE_TASK_ARG_LEN + 136)
-/* tse_task is used to track single asynchronous operation (8 bytes used for public members). */
+/** tse_task is used to track single asynchronous operation (8 bytes used for public members). */
 #define TSE_TASK_SIZE		(TSE_PRIV_SIZE + 8)
 
+/** A single task */
```
Comment (mchaarawi) [inline src/include/daos/tse.h:29]: that sounds like a useless comment.

File: src/include/daos_api.h [hunk 7]
```diff
@@ -16,6 +16,8 @@
 extern "C" {
 #endif
 
+#include <daos_types.h>
+
```
Comment (mchaarawi) [inline src/include/daos_api.h:20]: so even though those headers are installed, i believe we discussed at one point that really the user should just include daos.h for the daos API, then add daos_fs.h, or others for the higher level APIs vs being able to add each header. so your changes to make each public header byitself pull the other needed ones is not necessary.

**Review by mchaarawi — CHANGES_REQUESTED** [review]

File: src/include/daos/dpar.h (hunk 4 above)
Comment (ashleypittman) [reply src/include/daos/dpar.h:12]: This is an internal header file so doesn't need c++ protection.

tse.h however does not have these guards in but I'm not sure if it's worth adding them.

File: src/include/daos/tse.h (hunk 6 above)
Comment (ashleypittman) [reply src/include/daos/tse.h:29]: Our documentation requires all structures and entries to be documented, some text is needed here, feel free to suggest something better.

File: src/include/daos/mgmt.h [hunk 8]
```diff
@@ -10,9 +10,9 @@
 #ifndef __DC_MGMT_H__
 #define __DC_MGMT_H__
 
-#include <daos/common.h>
-#include <daos/tse.h>
```
Comment (daltonbohning) [inline src/include/daos/mgmt.h:14]: This seems improper, or at least not how I'm familiar with imports. Generally, if a file needs a definition, it should import the header that contains that definition.
For example, `mgmt.h` uses `tse_task_t`, which is defined in `tse.h`. So `mgmt.h` should include `tse.h` directly instead of implicitly relying on other imports to import those definitions, which could change.

File: src/include/daos/common.h [hunk 9]
```diff
@@ -25,14 +25,17 @@
...
 #include <daos_errno.h>
-#include <daos/debug.h>
-#include <gurt/hash.h>
-#include <gurt/common.h>
-#include <cart/api.h>
 #include <daos_types.h>
 #include <daos_obj.h>
 #include <daos_prop.h>
 #include <daos_security.h>
+#include <daos/tse.h>
```
Comment (daltonbohning) [inline src/include/daos/common.h:32]: `daos/common.h` includes `daos/tse.h`, but `daos/common.h` doesn't actually need the definitions in `daos/tse.h`. So this seems improper to me

File: src/include/daos/mgmt.h (hunk 8 above)
Comment (ashleypittman) [reply src/include/daos/mgmt.h:14]: You're correct.  I started this PR assuming that tse.h was a private header and trying to remove it's special handing but as @mchaarawi pointed it it's part of the API and it's special handing is warranted.  An early change I made incorrectly meant that daos_task.h neeed to included before daos/tse.h and these changes were to support that.

What you're saying about includes only being included to satisfy the header they're included in is good practice and this PR is an attempt to improve on that for the public headers but we do tend to have a few header files that just catch everything and I didn't mean to exacerbate that.

I've backed those changes out now and this PR is smaller as a result, thank you.

Comment (ashleypittman) [reply src/include/daos/mgmt.h:14]: The other issue here is that we should also be using <> for public headers and "" for internal ones, plus including system headers first, then daos public, then daos private but I don't think any of our code does that.

We've also got a few headers which are included both with or without a directory in places, daos_test.h is included as "<daos_test>" 25 times, <daos_test.h> 6 times and "<suite/daos_test.h>" once.

File: src/include/daos/container.h [hunk 10]
```diff
@@ -1,5 +1,5 @@
 /**
- * (C) Copyright 2016-2023 Intel Corporation.
+ * (C) Copyright 2016-2024 Intel Corporation.
```
Comment (daltonbohning) [inline src/include/daos/container.h:2]: erroneous copyright update after changes were backed out. Here and others

File: src/include/daos/pool.h [hunk 11]
```diff
@@ -12,13 +12,15 @@
...
+#include <daos_pool.h>
+#include <daos_task.h>
+
 #include <daos/common.h>
 #include <gurt/hash.h>
 #include <daos/pool_map.h>
 #include <daos/rsvc.h>
-#include <daos/tse.h>
-#include <daos_types.h>
-#include <daos_pool.h>
```
Comment (daltonbohning) [inline src/include/daos/pool.h:21]: This file should still `include <daos/tse.h>` right?

File: src/include/daos/task.h [hunk 12]
```diff
@@ -12,7 +12,6 @@
 
 #include <daos_types.h>
 #include <daos_api.h>
-#include <daos/tse.h>
```
Comment (daltonbohning) [inline src/include/daos/task.h:15]: And this one?

File: src/object/obj_internal.h [hunk 13]
```diff
@@ -14,7 +14,6 @@
 #include <stdint.h>
 #include <daos/common.h>
 #include <daos/event.h>
-#include <daos/tse.h>
```
Comment (daltonbohning) [inline src/object/obj_internal.h:17]: ditto

**Review by daltonbohning — COMMENTED** [review]
Please do correct me if I'm wrong about how includes should be organized

File: src/include/daos/pool.h (hunk 11 above)
Comment (ashleypittman) [reply src/include/daos/pool.h:21]: On what basis?  That in uses some of the defines in daos/tse.h I assume?  We have to assume that every file that includes this already includes daos/tse.h via some route.

It's easy enough to run my new unit test on the private headers as well, I'll update this PR with a few more code changes to allow all of them to be included but I won't enable this by default as it means the test would need to access abt.h and also we have daos/qat.h which we don't seem to include from anywhere today.

I've also found that using the macros from gurt/debug.h depends on gurt/common.h but you can't include gurt/common.h before gurt/debug.h so you have to include gurt/debug.h then gurt/common.h or just gurt/common.h

File: src/include/daos/container.h (hunk 10 above)
Comment (ashleypittman) [reply src/include/daos/container.h:2]: I'll do another pass and fix these when I'm done with other changes.

File: src/include/daos/pool.h (hunk 11 above)
Comment (daltonbohning) [reply src/include/daos/pool.h:21]: > On what basis? That in uses some of the defines in daos/tse.h I assume?

Right.

>  We have to assume that every file that includes this already includes daos/tse.h via some route.

That works accidentally, but not in the general sense. `daos/pool.h` now has a silent dependency on `daos/tse.h`.


> the macros from gurt/debug.h depends on gurt/common.h

That issue is exactly the kind of issue I'm trying to avoid. If "the macros from gurt/debug.h depends on gurt/common.h" then `gurt/debug.h` should include `gurt/common.h`

Comment (ashleypittman) [reply src/include/daos/pool.h:21]: > > On what basis? That in uses some of the defines in daos/tse.h I assume?
> 
> Right.
> 
> > We have to assume that every file that includes this already includes daos/tse.h via some route.
> 
> That works accidentally, but not in the general sense. `daos/pool.h` now has a silent dependency on `daos/tse.h`.

Having testing more now I don't think it does, `daos/pool.h` obviously pulls in `daos/tse.h` from one of the header files that it includes.  We could no doubt do some pruning of includes from c files but getting the headers right is the first step towards that.

> > the macros from gurt/debug.h depends on gurt/common.h
> 
> That issue is exactly the kind of issue I'm trying to avoid. If "the macros from gurt/debug.h depends on gurt/common.h" then `gurt/debug.h` should include `gurt/common.h`

It can't, `gurt/common.h` includes `gurt/debug.h`.  I'll deal with that another day.

Comment (daltonbohning) [reply src/include/daos/pool.h:21]: This is what I mean. Currently, `pool.h` has an _**implicit**_ dependency on `tse.h`. If, for whatever reason, `something.h` gets refactored and doesn't need `tse.h` anymore, then `pool.h` will be missing an include. And since `pool.h` already includes `tse.h`, I don't see the value in removing it.
include

Comment (daltonbohning) [reply src/include/daos/pool.h:21]: > We could no doubt do some pruning of includes from c files

I don't think the goal should be to reduce the includes to the minimal possible that allows us to build each header independently. We should include what we need in the headers that need it. Otherwise, we end up breaking unrelated files if we reorganize, like in my example above.

File: src/include/daos/common.h [hunk 14]
```diff
@@ -1,5 +1,5 @@
 /**
```
Comment (mchaarawi) [inline src/include/daos/common.h:1]: not quite sure why the change is needed in this file. could you explain?

File: src/include/daos/mgmt.h [hunk 15]
```diff
@@ -10,9 +10,10 @@
...
 #include <daos_types.h>
+#include <daos/tse.h>
+
+#include <daos/common.h>
```
Comment (mchaarawi) [inline src/include/daos/mgmt.h:16]: the change here just seems to remove and re-add the header with a line in between? just revert this so this file is not changed in the PR then.

File: src/include/daos/sys_debug.h [hunk 16]
```diff
@@ -10,9 +10,7 @@
 #ifndef __DC_DEBUG_H__
 #define __DC_DEBUG_H__
 
-#include <daos/common.h>
-#include <daos/tse.h>
-#include <daos_types.h>
+#include <daos_task.h>
```
Comment (mchaarawi) [inline src/include/daos/sys_debug.h:13]: so why here add daos_task.h but other places like https://github.com/daos-stack/daos/pull/13613/files#diff-930d69e66aac2e78ef8537ea29d976fc69620cbdeee43d117662b9e6f7f1f036R14 you added aos/tse.h

seems inconsistent.

File: src/include/daos/tse.h (hunk 6 above)
Comment (mchaarawi) [reply src/include/daos/tse.h:29]: "struct for a task object that is used for tracking an operation"

File: src/include/daos/tse.h (hunk 5 above)
Comment (mchaarawi) [reply src/include/daos/tse.h:31]: "result of the task operation - valid after task completion"

File: src/include/daos/tse.h [hunk 17]
```diff
@@ -36,37 +42,41 @@ typedef struct tse_task {
  * Track all of the tasks under a scheduler.
  */
 typedef struct {
+	/** the result */
```
Comment (mchaarawi) [inline src/include/daos/tse.h:45]: same as above, but this is for the schedule

File: src/include/daos/tse.h [hunk 18]
```diff
@@ -36,37 +42,41 @@ typedef struct tse_task {
...
-	/* daos schedule internal */
+	/** daos schedule internal */
 	struct {
 		uint64_t	ds_space[48];
 	}			ds_private;
 } tse_sched_t;
 
+/** Callaback function for when a scheduler is done */
 typedef int (*tse_sched_comp_cb_t)(void *args, int rc);
+/** Callaback function for when a task is done */
```
Comment (mchaarawi) [inline src/include/daos/tse.h:59]: this comment is wrong.

File: src/include/daos_array.h [hunk 19]
```diff
@@ -19,6 +19,10 @@
 extern "C" {
 #endif
 
+#include <daos_types.h>
+#include <daos_obj.h>
+#include <daos_obj_class.h>
+
```
Comment (mchaarawi) [inline src/include/daos_array.h:25]: im not going to comment again on every one of those public headers and will just comment on his one.
so the idea is for user to include daos.h, and that pulls in all what's needed. so "fixing" all the headers to be able to be included by itself is not required.

File: src/include/daos_fs.h [hunk 20]
```diff
@@ -21,8 +21,15 @@ extern "C" {
...
 #include <dirent.h>
+#include <inttypes.h>
 #include <sys/stat.h>
 
+#include <daos_types.h>
+#include <daos_obj.h>
+#include <daos_obj_class.h>
+#include <daos_array.h>
+#include <daos_cont.h>
+
```
Comment (mchaarawi) [inline src/include/daos_fs.h:32]: not required. expectation is that user pulls in daos.h then daos_fs.h

**Review by mchaarawi — CHANGES_REQUESTED** [review]

**Review by mchaarawi — APPROVED** [review]

File: utils/utest.yaml [hunk 21]
```diff
@@ -1,6 +1,9 @@
-#  (C) Copyright 2023 Intel Corporation.
+#  (C) Copyright 2023-2024 Intel Corporation.
 #
 #  SPDX-License-Identifier: BSD-2-Clause-Patent
+- name: include checks
+  tests:
+    - cmd: ["utils/include_test.py"]
```
Comment (daltonbohning) [inline utils/utest.yaml:6]: This is another case where my might want a directory of pytest tests somewhere

**Review by daltonbohning — APPROVED** [review]
